### PR TITLE
add xgboost

### PR DESCRIPTION
**Pull request by rootfs**
fix CI issue
https://github.com/sustainable-computing-io/kepler/pull/1052

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -26,6 +26,18 @@ jobs:
       run: |
         ./verify.sh ${{matrix.ebpf_provider}}
 
+  test_xgboost:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v4
+    - uses: ./
```
Comment (jiere) [inline .github/workflows/test.yml:33]: Do we really need this line? Similar question to other places in this workflow yaml.

Comment (rootfs) [reply .github/workflows/test.yml:33]: good question :D i copied it from other actions

File: index.js [hunk 2]
```diff
@@ -30,6 +30,36 @@ async function bcc() {
   return
 }
 
+async function xgboost() {
+  let artifacts_version = core.getInput('artifacts_version');
+  core.debug(artifacts_version);
+  if (artifacts_version === undefined || artifacts_version == null || artifacts_version.length === 0) {
```
Comment (vprashar2929) [inline index.js:36]: Can we convert this to a function? Something like:
```
const DEFAULT_ARTIFACTS_VERSION="0.26.0";
const getInputOrDefault = (input, defaultValue) => {
   if (input === undefined || input === null){
      return defaultValue;
   }
   return input;
};
...
let artifacts_version = getInputOrDefault(core.getInput('artifacts_version'),DEFAULT_ARTIFACTS_VERSION);
```

File: index.js [hunk 3]
```diff
@@ -30,6 +30,36 @@ async function bcc() {
...
+    artifacts_version="0.26.0";
+  }
+  let xgboost_version = core.getInput('xgboost_version');
+  core.debug(xgboost_version);
+  if (xgboost_version === undefined || xgboost_version == null || xgboost_version.length === 0) {
+    xgboost_version="2.0.1";
+  }
+  core.info(`Get xgboost with version:` + xgboost_version);
+  if (shell.exec("wget https://github.com/sustainable-computing-io/kepler-ci-artifacts/releases/download/v"+artifacts_version+"/xgboost-"+xgboost_version+"-Linux.sh.tar.gz").code !==0){
```
Comment (vprashar2929) [inline index.js:45]: Same here as well like:
```
const executeShell = (command, failMessage) => {
   if (shell.exec(command).code !=0) {
     shell.echo(`fail to get `+ failMessage);
     shell.exit(1);
   }
};
...
executeShell(`wget https://github.com/sustainable-computing-io/kepler-ci-artifacts/releases/download/v"+artifacts_version+"/xgboost-"+xgboost_version+"-Linux.sh.tar.gz`,"get xgboost pkg");
...
```

File: index.js (hunk 2 above)
Comment (rootfs) [reply index.js]: done

File: index.js (hunk 3 above)
Comment (rootfs) [reply index.js]: done

**Comment by rootfs** [comment]
@vprashar2929 good points! code refactored, let's see how the CI works

File: .github/workflows/test.yml (hunk 1 above)
Comment (jiere) [reply .github/workflows/test.yml:33]: Try remove them and see what happens, I did not see such usage in other repo GHAs

Comment (SamYuan1990) [reply .github/workflows/test.yml:33]: I suppose we will need this line for invoke GHA, @rootfs, @jiere, please take a look at github action official guidance at https://github.com/actions/javascript-action/blob/main/.github/workflows/ci.yml#L54

File: index.js [hunk 4]
```diff
@@ -1,122 +1,110 @@
...
-  if (shell.exec("sudo dpkg -i libbcc_"+bcc_version+"-1_amd64.deb").code !== 0) {
-    shell.echo("fail to install BCC deb");
+  return input;
+}
+
+function executeCommand(command, errorMessage) {
+  const result = shell.exec(command);
+  if (result.code !== 0) {
+    shell.echo(errorMessage);
     shell.exit(1);
```
Comment (SamYuan1990) [inline index.js:16]: maybe `shell.exit(result.code)` is better?

File: index.js [hunk 5]
```diff
@@ -1,122 +1,110 @@
...
+}
+
+function installBcc(bcc_version) {
+  core.info(`Get BCC with version:` + bcc_version);
+  executeCommand("wget https://github.com/sustainable-computing-io/kepler-ci-artifacts/releases/download/v"+bcc_version+"/bcc_v"+bcc_version+".tar.gz", "fail to get BCC deb");
+  executeCommand("tar -zxvf bcc_v"+bcc_version+".tar.gz", "fail to untar BCC deb");
+  executeCommand("sudo dpkg -i libbcc_"+bcc_version+"-1_amd64.deb", "fail to install BCC deb");
+}
+
+function installXgboost(artifacts_version, xgboost_version) {
```
Comment (SamYuan1990) [inline index.js:32]: may I know if Xgboost is available on GHA default runner?

File: index.js [hunk 6]
```diff
@@ -1,122 +1,110 @@
...
-    cluster_provider="kind";
-  }
-  let local_dev_cluster_version = core.getInput('local_dev_cluster_version');
-  core.debug(local_dev_cluster_version);
-  if (local_dev_cluster_version === undefined || local_dev_cluster_version == null || local_dev_cluster_version.length === 0) {
-    local_dev_cluster_version="main";
-  }
+  const cluster_provider = getInputOrDefault('cluster_provider', 'kind');
+  const local_dev_cluster_version = getInputOrDefault('local_dev_cluster_version', 'main');
+  const prometheus_enable = getInputOrDefault('prometheus_enable', '');
```
Comment (SamYuan1990) [inline index.js:53]: `''` or `false` ?

File: index.js [hunk 7]
```diff
@@ -1,122 +1,110 @@
...
+  const prometheus_operator_version = getInputOrDefault('prometheus_operator_version', '');
+  const grafana_enable = getInputOrDefault('grafana_enable', '');
+
   core.info(`Get local-cluster-dev with version `+ local_dev_cluster_version);
   shell.exec("git clone -b "+local_dev_cluster_version+" https://github.com/sustainable-computing-io/local-dev-cluster.git --depth=1");
```
Comment (SamYuan1990) [inline index.js:58]: change everything to `executeCommand` ?

File: index.js [hunk 8]
```diff
@@ -1,122 +1,110 @@
...
-  if (grafana_enable !== undefined && grafana_enable!=null && grafana_enable.length!=0) {
+  if (grafana_enable.length !== 0) {
     core.info(`use grafana enable `+grafana_enable);
-    //   grafana_enable, GRAFANA_ENABLE
     parameterExport = parameterExport + " && export GRAFANA_ENABLE="+grafana_enable;
   }
+
   parameterExport = parameterExport + " && "
   core.debug(parameterExport);
   shell.exec(parameterExport +` cd local-dev-cluster && bash -c './main.sh up'`)
```
Comment (SamYuan1990) [inline index.js:77]: change everything to executeCommand ?

File: index.js [hunk 9]
```diff
@@ -1,122 +1,110 @@
...
-  core.debug(cluster_provider);
-  const ebpfprovider = core.getInput('ebpfprovider');
+  const runningBranch = getInputOrDefault('runningBranch', '');
+  const ebpfprovider = getInputOrDefault('ebpfprovider', '');
+
   try {
+    // always install xgboost
+    const artifacts_version = getInputOrDefault('artifacts_version', '0.26.0');
+    const xgboost_version = getInputOrDefault('xgboost_version', '2.0.1');
+    installXgboost(artifacts_version, xgboost_version);
```
Comment (SamYuan1990) [inline index.js:88]: we should have a control flag for installXgboost

**Review by SamYuan1990 — CHANGES_REQUESTED** [review]
1. as we create a new function as `executeCommand`, we should better put every command running with it. and I am now further considering that we put the commands we are going to run in to a queue and then a for loop to run the commands instead `executeCommand` every where.(updates, for the queue based refactor, we can have another PR as this PR is major supports for `executeCommand`)
2. is xgboot should be installed as linux header and it works for GHA provided default runner? 
- if not, a control flag as bcc or cluster provider must been set in this PR.
- if yes or when we enable xgboot, it plays a role like kubectl or kind binary.... then, at cluster set up phase, it should move to cluster set up phase as InstallKubectl. and At verify phase, it should be verified together with cluster info?
3. some of the install functions running with await and some not, as this PR changes async way... we should keep our code in same style either await or async.

File: index.js (hunk 5 above)
Comment (rootfs) [reply index.js:32]: yes, xgboost is needed in default runner

File: index.js (hunk 4 above)
Comment (rootfs) [reply index.js:16]: this is to keep the consistency with previous code. If error occurs, the message is sufficient for debugging.

File: index.js (hunk 6 above)
Comment (rootfs) [reply index.js:53]: 'false' would be treated as a version string

**Comment by rootfs** [comment]
@SamYuan1990 refactored.

File: index.js (hunk 4 above)
Comment (SamYuan1990) [reply index.js]: for debugging usage, we can enable debugging flag at GHA web page UI... ok, let's keep current design.

**Comment by SamYuan1990** [comment]
> @SamYuan1990 refactored.

emmmm, ref 
https://github.com/sustainable-computing-io/kepler-action/pull/85/files#r1398214261

> yes, xgboost is needed in default runner

@rootfs , it's late my local time and seems you are WIP with code refactor.... may I have some questions and I am going to review this PR later on.
or @jiere can make a call to me on Monday or Tue?
We need further knowledge to review this PR. How does it work for kepler? Specially when we deploy kepler on a k8s node. How do we make kepler inside container to use this package on the k8s node?
- if it running as ebpf by "mount" which means we should make this function to ebpf part(as linux header).
- of it just like ebpf for VM or BM... for VM maybe we don't have xgboost supported? hence it more like ebpf or current code branch, then it LGTM.
- if it running ... for each cluster no matter ebpf support or not, as kubectl or just before we start cluster as `local-dev cluster`?

**Comment by rootfs** [comment]
@SamYuan1990 I hear you. The xgboost installation can be skipped if no version string is provided.

File: .github/workflows/test.yml (hunk 1 above)
Comment (jiere) [reply .github/workflows/test.yml:33]: Thanks for pointing out this, but I am still confused for such usage, can you find GHA document where mentioned this? I do realize that the `kepler-action` repo was generated from template of `actions/javascript-action`

**Comment by rootfs** [comment]
any more comment? want to get a new GHA soon.

File: .github/workflows/test.yml (hunk 1 above)
Comment (SamYuan1990) [reply .github/workflows/test.yml:33]: IMO it's just to invoke current version of GHA to test.

**Review by SamYuan1990 — APPROVED** [review]
LGTM

**Pull request merged**